src/FsAutoComplete.Core/UntypedAstUtils.fs

```diff
@@ -1,668 +1,663 @@
-    List.iter walkSynModuleOrNamespace moduleOrNamespaceList
-
-  and walkSynModuleOrNamespace (SynModuleOrNamespace(decls = decls; attribs = AllAttrs attrs; range = r)) =
-    addIfInside r
-    List.iter walkAttribute attrs
-    List.iter walkSynModuleDecl decls
-
-  and walkAttribute (attr: SynAttribute) =
-    addIfInside attr.Range
-    walkExpr attr.ArgExpr
-
-  and walkTyparDecl (SynTyparDecl(attributes = AllAttrs attrs; Item2 = typar)) =
-    List.iter walkAttribute attrs
-    walkTypar typar
-
-  and walkTyparDecls (typars: SynTyparDecls) =
-    typars.TyparDecls |> List.iter walkTyparDecl
-    typars.Constraints |> List.iter walkTypeConstraint
-
-  and walkSynValTyparDecls (SynValTyparDecls(typars, _)) = Option.iter walkTyparDecls typars
-
-  and walkTypeConstraint =
+  /// An recursive pattern that collect all sequential expressions to avoid StackOverflowException
+  let rec (|Sequentials|_|) =
     function
-    | SynTypeConstraint.WhereTyparIsValueType(t, r)
-    | SynTypeConstraint.WhereTyparIsReferenceType(t, r)
-    | SynTypeConstraint.WhereTyparIsUnmanaged(t, r)
-    | SynTypeConstraint.WhereTyparSupportsNull(t, r)
-    | SynTypeConstraint.WhereTyparIsComparable(t, r)
-    | SynTypeConstraint.WhereTyparIsEquatable(t, r) ->
-      addIfInside r
-      walkTypar t
-    | SynTypeConstraint.WhereTyparDefaultsToType(t, ty, r)
-    | SynTypeConstraint.WhereTyparSubtypeOfType(t, ty, r) ->
-      addIfInside r
-      walkTypar t
-      walkType ty
-    | SynTypeConstraint.WhereTyparIsEnum(t, ts, r)
-    | SynTypeConstraint.WhereTyparIsDelegate(t, ts, r) ->
-      addIfInside r
-      walkTypar t
-      List.iter walkType ts
-    | SynTypeConstraint.WhereTyparSupportsMember(t, sign, r) ->
-      addIfInside r
-      walkType t
-      walkMemberSig sign
-    | SynTypeConstraint.WhereSelfConstrained(t, r) ->
-      addIfInside r
-      walkType t
+    | SynExpr.Sequential(_, _, e, Sequentials es, _) -> Some(e :: es)
+    | SynExpr.Sequential(_, _, e1, e2, _) -> Some [ e1; e2 ]
+    | _ -> None
 
-  and walkPat =
+  let (|ConstructorPats|) =
     function
-    | SynPat.Tuple(elementPats = pats; range = r)
-    | SynPat.ArrayOrList(_, pats, r)
-    | SynPat.Ands(pats, r) ->
-      addIfInside r
-      List.iter walkPat pats
-    | SynPat.Named(ident, _, _, r) -> addIfInside r
-    | SynPat.Typed(pat, t, r) ->
-      addIfInside r
-      walkPat pat
-      walkType t
-    | SynPat.Attrib(pat, AllAttrs attrs, r) ->
-      addIfInside r
-      walkPat pat
+    | SynArgPats.Pats ps -> ps
+    | SynArgPats.NamePatPairs(pats = xs) -> xs |> List.map (fun (_, _, pat) -> pat)
+
+  /// A pattern that collects all patterns from a `SynSimplePats` into a single flat list
+  let (|AllSimplePats|) (pats: SynSimplePats) =
+    let rec loop acc pat =
+      match pat with
+      | SynSimplePats.SimplePats(pats = pats) -> acc @ pats
+
+    loop [] pats
+
+  type SyntaxCollectorBase() =
```

Comment:
To make this type I looked at every `walkX` member in the walker function from the previous version and made a `WalkX` member on this type. The walker function handles going 'deeper' into the structure, so the member  overrides that users write here should be strictly focused on looking at this particular syntax node, not doing traversal themselves.